cli/exchange/node.go

```diff
@@ -465,3 +480,341 @@ func NodeListErrors(org string, credToUse string, node string, long bool) {
 		fmt.Printf("%s\n", jsonBytes)
 	}
 }
+
+func ValidateUserInput(org string, credToUse string, userInput policy.UserInput, exNode ExchangeNodes) (bool, string, error) {
+	msgPrinter := i18n.GetMessagePrinter()
+	warningMessage := ""
+
+	serviceOrg := userInput.ServiceOrgid
+	serviceUrl := userInput.ServiceUrl
+	serviceVersionRange := userInput.ServiceVersionRange
+	serviceArch := userInput.ServiceArch
+
+	msgPrinter.Printf("Validating userinput %v, %v, %v, %v", serviceOrg, serviceUrl, serviceVersionRange, serviceArch)
+	msgPrinter.Println()
+
+	var errorString string
+	if &userInput.ServiceVersionRange == nil || userInput.ServiceVersionRange == "" {
+		def := "0.0.0"
+		serviceVersionRange = def
+	}
+
+	// Convert the version to a version expression.
+	vExp, err := semanticversion.Version_Expression_Factory(serviceVersionRange)
+	if err != nil {
+		errorString = msgPrinter.Sprintf("versionRange %v cannot be converted to a version expression, error %v", userInput.ServiceVersionRange, err)
+		return false, warningMessage, errors.New(errorString)
+	}
+	serviceVersion := vExp.Get_expression()
+
+	// service exist
+	var sdef *exchange.ServiceDefinition
+
+	searchVersion, err := getSearchVersion(serviceVersion)
+	if err != nil {
+		return false, warningMessage, err
+	}
+
+	sdef, err = getService(cliutils.OrgAndCreds(org, credToUse), serviceOrg, serviceUrl, serviceArch, serviceVersion, searchVersion)
+	if sdef == nil || err != nil {
+		return false, warningMessage, err
+	}
+
+	// compare ServiceDefinition.UserInput(array) with userInput.Inputs(array)
+	var compareResult bool
+	compareResult, err = compareUserInput(sdef.UserInputs, userInput.Inputs)
+	if !compareResult {
+		errorString = msgPrinter.Sprintf("Failed to compare given userInput with service userInput, error: %v", err)
+		return false, warningMessage, errors.New(errorString)
+	} else if err != nil {
+		// give back a warning
+		warningMessage = msgPrinter.Sprintf("For service %v %v %v, got warning message: %v", serviceOrg, serviceUrl, serviceArch, err.Error())
+	}
+
+	// if node has a pattern, check the service is defined in top level services
+	nodes := exNode.Nodes
+	if len(nodes) != 1 {
+		errorString = msgPrinter.Sprintf("Expecting 1 exchange node, but get %d nodes", len(nodes))
+		return false, warningMessage, errors.New(errorString)
+	}
+
+	nodePatternString := ""
+	for _, node := range nodes {
+		nodePatternString = node.Pattern
+	}
+
+	// Need to check:
+	// 1) if service is top level service
+	// 2) or recursively check if service defined in dependent service
+	// 3) if service range of the dependent service has intersection with versionRange of service from userInput
+	// 4) the required service has userinputs should have default values
```

Comment:
A node with a pattern, where the pattern's top level service requires another service that has userinputs but no default values, is not flagged as in error. Case 4 here does not appear to work correctly.